File: javaproject/src/main/java/ui/GraficoPizza.java
package ui;

import java.awt.Color;
import java.awt.Font;
import java.awt.Graphics;
import java.awt.Graphics2D;
import java.awt.RenderingHints;
import java.util.ArrayList;
import java.util.List;


public class GraficoPizza extends Grafico {

    private List<Double> valores;
    private String[] categorias;

    // -------------------------------------------------
    private final Color[] cores = {
        new Color(31, 119, 180),   // Azul
        new Color(255, 127, 14),   // Laranja
        new Color(44, 160, 44),    // Verde
        new Color(214, 39, 40),    // Vermelho
        new Color(148, 103, 189),  // Roxo
        new Color(140, 86, 75),    // Marrom
        new Color(227, 119, 194),  // Rosa
        new Color(127, 127, 127),  // Cinza
        new Color(188, 189, 34),   // Oliva
        new Color(23, 190, 207),   // Ciano
        new Color(174, 199, 232),  // Azul Claro
        new Color(255, 187, 120)   // Laranja Claro
    };
    // -------------------------------------------------
    
    public GraficoPizza() {
        super();
        this.valores = new ArrayList<>();
        this.categorias = new String[0];
    }
    // -------------------------------------------------

    public void atualizarDados(String[] categorias, List<Double> valores) {
        this.categorias = (categorias != null) ? categorias : new String[0];
        this.valores = (valores != null) ? valores : new ArrayList<>();
        repaint();
    }
    // -------------------------------------------------
    
    @Override
    protected void paintComponent(Graphics g) {
        super.paintComponent(g);
        Graphics2D g2d = (Graphics2D) g;
        g2d.setRenderingHint(RenderingHints.KEY_ANTIALIASING, RenderingHints.VALUE_ANTIALIAS_ON);

        if (valores == null || valores.isEmpty() || !temDadosValidos()) {
            g2d.setFont(new Font("Arial", Font.PLAIN, 14));
            g2d.setColor(Color.GRAY);
            g2d.drawString("Sem dados para exibir o gráfico.", 20, 30);
            return;
        }

        int areaGraficoLargura = (int) (getWidth() * 0.60);
        int areaLegendaX = areaGraficoLargura + 10;
        int diametro = Math.min(areaGraficoLargura, getHeight()) - 40;
        int x = (areaGraficoLargura - diametro) / 2;
        int y = (getHeight() - diametro) / 2;

        double total = 100.0;
        double anguloInicial = 90.0;

        // Desenha as fatias da pizza
        for (int i = 0; i < valores.size(); i++) {
            double valorFatia = valores.get(i);
            if (valorFatia <= 0) continue;

            double anguloArco = (valorFatia / total) * 360.0;
            g2d.setColor(cores[i % cores.length]);
            g2d.fillArc(x, y, diametro, diametro, (int) anguloInicial, -(int) Math.round(anguloArco));
            anguloInicial -= anguloArco;
        }

        // Desenha a legenda
        g2d.setFont(new Font("Arial", Font.PLAIN, 11));
        int legendaY = y;
        for (int i = 0; i < categorias.length; i++) {
            if (valores.get(i) > 0) {
                g2d.setColor(cores[i % cores.length]);
                g2d.fillRect(areaLegendaX, legendaY, 12, 12);
                g2d.setColor(Color.BLACK);
                String textoLegenda = String.format("%s: %.2f%%", categorias[i], valores.get(i));
                g2d.drawString(textoLegenda, areaLegendaX + 20, legendaY + 11);
                legendaY += 20;
            }
        }
    }

    private boolean temDadosValidos() {
        if (valores == null) return false;
        for (double val : valores) {
            if (val > 0) return true;
        }
        return false;
    }
}

File: pythonproject/ui/grafico_pizza.py
from PyQt6.QtWidgets import QWidget
from PyQt6.QtGui import QPainter, QColor, QFont
from PyQt6.QtCore import Qt
from typing import List

class GraficoPizza(QWidget):
    def __init__(self, parent=None):
        super().__init__(parent)
        self.valores: List[float] = []
        self.categorias: List[str] = []
        self.cores = [
            QColor(31, 119, 180),    # Azul
            QColor(255, 127, 14),    # Laranja
            QColor(44, 160, 44),     # Verde
            QColor(214, 39, 40),     # Vermelho
            QColor(148, 103, 189),   # Roxo
            QColor(140, 86, 75),     # Marrom
            QColor(227, 119, 194),   # Rosa
            QColor(127, 127, 127),   # Cinza
            QColor(188, 189, 34),    # Oliva
            QColor(23, 190, 207),    # Ciano
            QColor(174, 199, 232),   # Azul Claro
            QColor(255, 187, 120)    # Laranja Claro
        ]
        self.setMinimumSize(400, 300)
        
    def atualizar_dados(self, categorias: List[str], valores: List[float]):
        self.categorias = categorias.copy() if categorias else []
        self.valores = valores.copy() if valores else []
        self.update()
        
    def paintEvent(self, event):
        painter = QPainter(self)
        painter.setRenderHint(QPainter.RenderHint.Antialiasing)
        
        if not self.valores or not any(val > 0 for val in self.valores):
            painter.setFont(QFont("Arial", 10))
            painter.setPen(QColor(128, 128, 128))
            painter.drawText(20, 30, "Sem dados para exibir o gráfico.")
            return
            
        area_grafico_largura = int(self.width() * 0.60)
        area_legenda_x = area_grafico_largura + 10
        diametro = min(area_grafico_largura, self.height()) - 40
        x = (area_grafico_largura - diametro) // 2
        y = (self.height() - diametro) // 2
        
        total = 100.0
        angulo_inicial = 90.0

        for i, valor in enumerate(self.valores):
            if valor <= 0:
                continue
                
            angulo_arco = (valor / total) * 360.0
            painter.setBrush(self.cores[i % len(self.cores)])
            painter.setPen(QColor(0, 0, 0))
            painter.drawPie(x, y, diametro, diametro, int(angulo_inicial * 16), -int(angulo_arco * 16))
            angulo_inicial -= angulo_arco

        painter.setFont(QFont("Arial", 9))
        legenda_y = y
        
        for i, categoria in enumerate(self.categorias):
            if i < len(self.valores) and self.valores[i] > 0:
                painter.setBrush(self.cores[i % len(self.cores)])
                painter.drawRect(area_legenda_x, legenda_y, 12, 12)
                painter.setPen(QColor(0, 0, 0))
                texto_legenda = f"{categoria}: {self.valores[i]:.2f}%"
                painter.drawText(area_legenda_x + 20, legenda_y + 11, texto_legenda)
                legenda_y += 20